[Merged by Bors] - feat: ample sets in real vector spaces

grunweg opened the pull request:
From the sphere eversion project.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ADedecker commented on Mathlib/Analysis/Convex/AmpleSet.lean:
```diff
@@ -0,0 +1,94 @@
+
+/-- Images of ample sets under continuous affine equivalences are ample. -/
+theorem AmpleSet.image {s : Set E} (h : AmpleSet s) (L : E ≃ᵃL[ℝ] F) :
+    AmpleSet (L '' s) := fun x hx ↦ by
+  rw [L.image_eq_preimage] at hx
+  have : L '' connectedComponentIn s (L.symm x) = connectedComponentIn (L '' s) x := by
+    conv_rhs => rw [← L.apply_symm_apply x]
+    exact (L.toHomeomorph).image_connectedComponentIn hx
+  rw [← this]
+  refine (L.toAffineMap.image_convexHull _).symm.trans ?_
+  rw [h (L.symm x) hx, image_univ]
+  exact L.surjective.range_eq
```
```suggestion
    AmpleSet (L '' s) := forall_mem_image.mpr fun x hx ↦
  calc (convexHull ℝ) (connectedComponentIn (⇑L '' s) (L x))
    _ = (convexHull ℝ) (L '' (connectedComponentIn s x)) :=
          .symm <| congrArg _ <| L.toHomeomorph.image_connectedComponentIn hx
    _ = L '' (convexHull ℝ (connectedComponentIn s x)) :=
          .symm <| L.toAffineMap.image_convexHull _
    _ = univ := by rw [h x hx, image_univ, L.surjective.range_eq]
```

Looks a bit nicer I think?

grunweg replied:
Indeed!

ADedecker commented on Mathlib/Analysis/Convex/AmpleSet.lean:
```diff
@@ -0,0 +1,94 @@
+/-- The union of two ample sets is ample. -/
+theorem AmpleSet.union {s t : Set F} (hs : AmpleSet s) (ht : AmpleSet t) : AmpleSet (s ∪ t) := by
+  intro x hx
+  rcases hx with (h | h)
+  -- The connected component of `x ∈ s` in `s ∪ t` contains the connected component of `x` in `s`,
+  -- hence is also full; similarly for `t`.
+  · rw [← Set.univ_subset_iff, ← hs x h]
+    apply convexHull_mono
+    apply connectedComponentIn_mono
+    apply subset_union_left
+  · rw [← Set.univ_subset_iff, ← ht x h]
+    apply convexHull_mono
+    apply connectedComponentIn_mono
+    apply subset_union_right
```
In case where there is an obvious parallelism going on I like to do them simultaneously. Feel free to ignore if you don't like it.

```suggestion
  rcases hx with (h | h) <;>
  -- The connected component of `x ∈ s` in `s ∪ t` contains the connected component of `x` in `s`,
  -- hence is also full; similarly for `t`.
  [have hx := hs x h; have hx := ht x h] <;>
  rw [← Set.univ_subset_iff, ← hx] <;>
  apply convexHull_mono <;>
  apply connectedComponentIn_mono <;>
  [apply subset_union_left; apply subset_union_right]
```

grunweg replied:
I do (and didn't know that trick; TIL). Thanks for the suggestion.

ADedecker commented on Mathlib/Analysis/Convex/AmpleSet.lean:
```diff
@@ -0,0 +1,94 @@
+
+/-- The empty set in a vector space is ample. -/
+@[simp]
+theorem ampleSet_empty : AmpleSet (∅ : Set F) := fun _ h ↦ False.elim h
```
```suggestion
theorem ampleSet_empty : AmpleSet (∅ : Set F) := fun _ ↦ False.elim
```

grunweg commented:
I have incorporated your comments.
awaiting-review

ADedecker reviewed (approved):
LGTM, but obviously I don't want to merge this since I'm one of the authors of the original file (although I believe this part is mostly written by Floris).

fpvandoorn reviewed (approved):
LGTM

Maybe you can add `iff` versions of the last two lemmas?

I think three pairs of eyes is sufficient, even if we all contributed to the code in some way

bors d+

grunweg commented:
Good idea, added. Thanks for the reviews!
(I also added a few lemmas to ContinousAffineEquiv which hold for Equivs and are needed here. If this duplication can/should be reduced is a topic for another discussion, not today.)
bors r+
